### ISSUE TITLE
&& doesn't work

### ISSUE THREAD

**Issue by joshuaconner** [fish-shell/fish-shell#150]
`cd .. && pwd` returns the following error: 

```
fish: Expected a command name, got token of type 'Run job in background'. Did you mean 'COMMAND; and COMMAND'? See the help section for the 'and' builtin command by typing 'help and'.
cd .. && pwd
```

Obviously I can use the `COMMAND; and COMMAND` syntax, but it would be nice if fish supported this.

Thanks!

**Comment by rhoconlinux** [comment]
AWESOME :+1:

**Comment by ylluminate** [comment]
Any running changelog available?  It appears that the root `CHANGELOG` is quite out of date.

**Comment by Globegitter** [comment]
@pragmattica Now that fish 2.2 is out I tried that on ubuntu 14.04 and `echo 'test1' && echo 'test2';` turns into `echo test1 &; and  echo 'test2';`

Could there be an issue somewhere within the `fish_user_key_bindings.fish`? (I will try to look over it myself, but my fish/bash script knowledge is a bit shaky)

**Comment by pickfire** [comment]
-1 to weird symbols:
```fish
if [[ a = b ]]; and [[ b = a ]]; or [[ c = b ]]; then echo hello; and echo world; done
```
+1 to simple symbols:
```fish
if [[ a = b ]] && [[ b = a ]] || [[ c = b ]]; then echo hello && echo world; done
```

**Comment by Globegitter** [comment]
Aaahhh I think I know what the issue is @pragmattica the command `commandline` returns the buffer without any quotation-marks. Yep that seems to be it. So once https://github.com/fish-shell/fish-shell/issues/2210 is fixed the above script should be fixed as well.

**Comment by jonathan-annett** [comment]
the omission of && and || simply means fish lives up to its Monica
fuckin' inadequate shell

**Comment by tiagoboldt** [comment]
Is this still begin considered?

**Comment by aboy021** [comment]
+1
The symbols aren't weird, they're commonly used for logic in a wide number of programming languages.
Copy and paste compatibility is a very real argument, and one I support.

**Comment by krader1961** [comment]
I'm going to be the bad guy and bring this discussion to a close. This will not be implemented. I'm especially bothered by the repeated statements about wanting to be able to "cut and paste commands from bash scripts". Even if fish supported `&&` and `||` operators that would not ensure you could simply cut and paste statements containing those operators. We do not want to give the impression that you can do so. Honestly, if you want bash/zsh syntax and behavior you should use those shells (or an alternative that claims compatibility with them).

Most of the other ideas discussed should have their own issues. In particular there may be some merit to implementing some of the `[[ ... ]]` behaviors that were first introduced by ksh88 (not bash). But probably via enhancing the builtin `test` command or a new command rather than by implementing those special tokens.

**Comment by ylluminate** [comment]
I am still very disturbed by this notion of not supporting zsh and bash.  I can understand the logic, but I don't agree with it.  I was very interested in fish some time ago, but moved back away due to this opposition.  There may be some opportunity to fork the project here if there are enough of us, as developers, who would like to see this implemented.  Sad to see this continued resistance which ultimately impedes the forward momentum and acceptance of such an interesting piece of work.

**Comment by ridiculousfish** [comment]
@ylluminate what do you have in mind by "supporting zsh and bash?" What level of compatibility would you like to see?

**Comment by krader1961** [comment]
> I am still very disturbed by this notion of not supporting zsh and bash. I can understand the logic...

I'm afraid I don't understand your logic for having fish support bash and zsh features. If you want Java why use C++? If you want Perl why use Python? As @ridiculousfish just asked, besides some syntactic sugar what else would fish have to support to make you happy? At what point does fish just become another bash clone?

**Comment by pickfire** [comment]
@krader1961 Nice speech, I wouldn't do scripting in fish all the time as (followed by orders):

1. fish's script is slow (slower than mksh)
2. some weird syntax (eg. `; and`, `; or` which looks ugly for me)
3. not posix compatible (I can say the syntax is easy but still need time to learn)

Well, I wouldn't care about the 2 and 3 only if fish script is **faster than other shells** (eg. mksh, bash).

On the other hand, I use fish because:

1. startup time is faster than zsh (but slower than bash)
2. have a lot of features enabled by default (syntax highlight is the best)
3. configurations is nice and simple (especially `set -Ux`)

**Comment by jakwings** [comment]
@ylluminate Looking at this four-year-old issue, I can see how lazy/busy the people who want to enjoy fish like bash, are still lazy/busy. The eager ones have been making their own shell (such as magicant/yash, elvish/elvish, michaelmacinnis/oh, and me :-P). If you want to do things in an un-fishy way, you probably need an un-fishy shell.

If you really consider how to implement `&& ||` onto fish and make that not very un-fishy, then you may see that you are just expelling `begin; and` (see what commit 594b460ba2d8dca59a3bfd282397c5f33aa9da6f do, counter-`;`, very ugly) and get only some little sugar.

**Comment by ridiculousfish** [comment]
@pickfire Can you elaborate on "fish's script is slow?" Are there any tests that lead you to that conclusion?

FWIW in my tests fish is faster than bash overall, by virtue of `posix_spawn` and its fast parser, though it may be slower in certain cases like `alias`.

**Comment by pickfire** [comment]
```
ivan@alarmpi /tmp> echo 'echo test' > script
ivan@alarmpi /tmp> time bash script
test
0.02user 0.01system 0:00.03elapsed 96%CPU (0avgtext+0avgdata 2776maxresident)k
0inputs+0outputs (0major+149minor)pagefaults 0swaps
ivan@alarmpi /tmp> time mksh script
test
0.00user 0.00system 0:00.02elapsed 0%CPU (0avgtext+0avgdata 1420maxresident)k
480inputs+0outputs (2major+82minor)pagefaults 0swaps
ivan@alarmpi /tmp> time fish script
test
0.07user 0.01system 0:00.09elapsed 85%CPU (0avgtext+0avgdata 4204maxresident)k
352inputs+0outputs (2major+231minor)pagefaults 0swaps
```

@ridiculousfish Just imagine it is already very slow compare to the others for just a single `echo`.

**Comment by Globegitter** [comment]
@krader1961 @ridiculousfish IMO, once https://github.com/fish-shell/fish-shell/issues/2210 gets released this can be closed. There is a very easy solution that @kballard posted.

**Comment by ridiculousfish** [comment]
@pickfire Thanks for sharing that. That test is measuring startup overhead, not the time of `echo`. Startup time is very important, but you can't draw conclusions beyond startup time (which is mostly dependent on config.fish) from that test.

Here's a different micro benchmark:

```
> cat test.fish
for i in (seq 1000)
	ls > /dev/null
end

> time fish test.fish
        1.51 real         0.74 user         0.65 sys
> cat test.sh
for i in {1..1000} ; do
	ls > /dev/null
done

> time bash test.sh
        2.01 real         0.85 user         1.12 sys
```

here fish wins by a lot, but this is mainly measuring the difference of fork vs posix_spawn.

I think one conclusion here is that we need a comprehensive benchmark suite.

**Comment by pickfire** [comment]
@ridiculousfish Here is it, the benchmark that you have gave me:
```fish
> cat test.fish
for i in (seq 1000)
    ls > /dev/null
end
> time fish test.fish
4.18user 4.04system 0:22.62elapsed 36%CPU (0avgtext+0avgdata 4360maxresident)k
96inputs+0outputs (1major+321041minor)pagefaults 0swaps
> cat test.sh
for i in {1..1000} ; do
    ls > /dev/null
done
> time bash test.sh
0.70user 1.62system 0:09.81elapsed 23%CPU (0avgtext+0avgdata 2844maxresident)k
0inputs+0outputs (0major+154100minor)pagefaults 0swaps
> time mksh test.sh
0.00user 0.01system 0:00.04elapsed 24%CPU (0avgtext+0avgdata 1780maxresident)k
752inputs+0outputs (3major+203minor)pagefaults 0swaps
```
The `mksh` seems to have an impressive speed.

**Comment by ridiculousfish** [comment]
@pickfire fish's `ls` is a function on Linux that does some other stuff to make the output nicer, which bash is not doing. If you want an apples-to-apples comparison on Linux, use `command ls`. (On Darwin, fish's `ls` wrapper is much thinner - I should have pointed that out but I forgot.)

The explanation for mksh's "impressive speed" in your test that `test.sh` script is not a valid mksh script, so it's not invoking `ls` at all. It's just outputting an error that gets sent to /dev/null.

This shows that performance measurement is very tricky - it's very easy to measure something different than what you think you're measuring!

**Comment by pickfire** [comment]
@ridiculousfish, is there any reason to use the `ls` function is fish script compare to `command ls`? The output isn't important during scripting.

Okay, this time I did what you told:
```fish
> cat test.fish
for i in (seq 1000)
    command ls > /dev/null
end
> time fish test.fish
0.66user 1.04system 0:08.08elapsed 21%CPU (0avgtext+0avgdata 4364maxresident)k
624inputs+0outputs (4major+113176minor)pagefaults 0swaps
> cat test.sh
for i in $(seq 1000) ; do
    ls > /dev/null
done
> time mksh test.sh
0.21user 0.65system 0:07.64elapsed 11%CPU (0avgtext+0avgdata 1884maxresident)k
0inputs+0outputs (0major+119632minor)pagefaults 0swaps
```
At the end, `mksh` is the fastest, now I find it impressive for `fish` to be faster than `mksh`. If `fish` has built-in parallel command execution support, I think it would have been the fastest of all.

**Comment by ridiculousfish** [comment]
That .66 user time is quite high. I speculate it's due to your config.fish contents - defining aliases and such. For the record, this is what I see on my Linux box (best out of 3 for each):

```
> time fish test.fish
0.15user 1.21system 0:01.33elapsed...
> time mksh test.mksh
0.12user 1.13system 0:01.24elapsed...
> time bash test.sh
0.18user 1.34system 0:01.47elapsed...
```

everyone is very close.

It might be interesting for any benchmark to incorporate a sort of "oracular shell" that just does a fork/exec or posix_spawn to some path command, so we can get a sense of how far we are from the maximum speed that the OS can deliver.

**Comment by pickfire** [comment]
@ridiculousfish, my config.fish is very short:
```fish
# Solarized colors
test $SHLVL = 2
and sh $HOME/.config/base16-shell/base16-solarized.dark.sh &
set fisher_home ~/.local/share/fisherman
set fisher_config ~/.config/fisherman
source $fisher_home/config.fish
```
Well probably it uses fisherman, not sure if it will slow it down. @bucaran is the nice guy for this. This time, I emptied the config.fish and see the results (best result of 5 retries):
```fish
> time fish test.fish; time mksh test.sh; time bash test.sh
0.62user 1.09system 0:07.28elapsed 23%CPU (0avgtext+0avgdata 4396maxresident)k
0inputs+0outputs (0major+108240minor)pagefaults 0swaps
0.21user 0.62system 0:06.92elapsed 11%CPU (0avgtext+0avgdata 1888maxresident)k
0inputs+0outputs (0major+116674minor)pagefaults 0swaps
0.29user 0.81system 0:07.78elapsed 14%CPU (0avgtext+0avgdata 2780maxresident)k
0inputs+0outputs (0major+145628minor)pagefaults 0swaps
```

**Comment by jakwings** [comment]
@ridiculousfish @pickfire Can you chat in another issue so I can block your messages?

**Comment by ridiculousfish** [comment]
Yes, sorry for spamming this issue. @pickfire if you'd like to follow up some benchmarks that you think are especially interesting or relevant, please feel free to open another issue.

**Comment by Scellow** [comment]
Can't we fucking have &&


Who the fuck prefer to write

'; and'  over '&&' ?

**Comment by Nodd** [comment]
About bash and zsh compatibility, I personally don't expect a full compatibility, but with every bit of compatibility added fish can become more popular.
Even without taking scripting into account, `&&` and `||` are important because they are very commonly used for chaining commands. As an example I still have problems with #2292. I know it can be considered as not a fish bug, but the fact is that improving fish compatibility would fix it.

**Comment by wilhelmy** [comment]
I'm with maxfl here, :-1: for having `||` and `&&` as part of the fish syntax, if this is supposed to be a popular vote.

**Comment by krader1961** [comment]
Fish is not bash (or zsh). There is no compelling reason to implement `&&` and `||` other than for compatibility with those shells. Compatibility with the status-quo is an anti-goal of this project.

@Nodd: Your reference to issue #2292 is actually an example of why fish should not implement that syntax. Doing so would be another step towards making people think fish is a POSIX (bash, zsh, etc.) clone. That would just create other demands that fish behave like those shells in other areas.

**Comment by Globegitter** [comment]
For all the people that are looking for a solution, since fish 2.3 I can confirm there is a working solution.

Just do what above answer mentions:

> Here you go. Once PR #1633 lands, you can take the following and save it in ~/.config/fish/functions/fish_user_key_bindings.fish:

```sh
function handle_input_bash_conditional --description 'Function used for binding to replace && and ||'
    # This function is expected to be called with a single argument of either & or |
    # The argument indicates which key was pressed to invoke this function
    if begin; commandline --search-mode; or commandline --paging-mode; end
        # search or paging mode; use normal behavior
        commandline -i $argv[1]
        return
    end
    # is our cursor positioned after a '&'/'|'?
    switch (commandline -c)[-1]
    case \*$argv[1]
        # experimentally, `commandline -t` only prints string-type tokens,
        # so it prints nothing for the background operator. We need -c as well
        # so if the cursor is after & in `&wat` it doesn't print "wat".
        if test -z (commandline -c -t)[-1]
            # Ideally we'd just emit a backspace and then insert the text
            # but injected readline functions run after any commandline modifications.
            # So instead we have to build the new commandline
            #
            # NB: We could really use some string manipulation operators and some basic math support.
            # The `math` function is actually a wrawpper around `bc` which is kind of terrible.
            # Instead we're going to use `expr`, which is a bit lighter-weight.

            # get the cursor position
            set -l count (commandline -C)
            # calculate count-1 and count+1 to give to `cut`
            set -l prefix (expr $count - 1)
            set -l suffix (expr $count + 1)
            # cut doesn't like 1-0 so we need to special-case that
            set -l cutlist 1-$prefix,$suffix-
            if test "$prefix" = 0
                set cutlist $suffix-
            end
            commandline (commandline | cut -c $cutlist)
            commandline -C $prefix
            if test $argv[1] = '&'
                commandline -i '; and '
            else
                commandline -i '; or '
            end
            return
        end
    end
    # no special behavior, insert the character
    commandline -i $argv[1]
end

function fish_user_key_bindings
    bind \& 'handle_input_bash_conditional \&'
    bind \| 'handle_input_bash_conditional \|'
end
```

**Comment by brj** [comment]
@Globegitter 

What about just?

```fish
function sudobangbang --on-event fish_postexec
    abbr !! sudo $argv[1]
end
```

**Comment by Globegitter** [comment]
@brj Also putting that into `~/.config/fish/functions/fish_user_key_bindings.fish`? That did not seem to do the trick for me.

**Comment by brj** [comment]
You need to source this function. So, your options are to put it in your config.fish or put it inside ~/.conf.d/[name-dont-matter].fish.

**Comment by faho** [comment]
@brj: Wrong issue for that function. This is about "&&" and "||", not "!!" - different things.

**Comment by brj** [comment]
@faho lol, you are right. I always get a notification on that one, so I thought this was another case of sudo !!. 

@Globegitter Sorry!

**Comment by gladhorn** [comment]
How about shipping the proposed work-around with fish as example, so users can conveniently copy it? The error message could easily be augmented to mention this. I keep on pasting the same one liner (from gerrit code-review) where && is used because it works in pretty much all shells (including windows cmd.exe besides unix-y shells). To me the above mentioned script is a productivity booster (I was typing bash -c "paste here" for a while...).

**Comment by kozec** [comment]
Same problem as above, I really can't find any way to do

`gcc test2.c -o a.out &&  ./a.out`

in fish.

**Comment by ivan** [comment]
```fish
gcc test2.c -o a.out; and ./a.out
```

**Comment by kozec** [comment]
@ivan tried that, but it gives "and" and "./a.out" as arguments to gcc.

**Comment by ivan** [comment]
I do this "completely impossible" thing quite often.  Maybe try the `; and` syntax in a clean fish environment?  If it doesn't, maybe something is up with your semicolons?  Are you typing this into fish, or is it getting passed it from another program?

**Comment by kozec** [comment]
@ivan Typing it in.

Sorry for probably stupid question, but how do I do parentheses? I can't find a word about them in documentation and it's not exactly googleable topic :(

**Comment by floam** [comment]
I think you're just having trouble grouping your items together here. What you need to use are "begin" like a left brace (that's how I think of it) and "end" like a right one.

**Comment by kozec** [comment]
@floam Thanks, that almost works. But stuff before & is not sent to background :(

Minimal example that should work anywhere:

`begin sleep 2s ; env DISPLAY=:2 xterm ; end & Xephyr :2`

sleeps 2s, then launches xterm, *then* starts XServer for that xterm. xterm ofc crashes before XServer becames ready :(

**Comment by floam** [comment]
Ah, sorry. The problem: https://github.com/fish-shell/fish-shell/issues/238

**Comment by floam** [comment]
You will find you can work around it by this where you must, not pretty:

`fish -c 'sleep 2s; env DISPLAY=:2 xterm' & Xephyr :2`

**Comment by kribblo** [comment]
When you have a large amount of existing scripts, for development, building and inside npm run-scripts it's not reasonable to switch, especially since some/most are not on fish. For the regular bash scripts with shebang and all it's not a problem of course.

The nice (or "nice") thing about && is that it works on the regular shells on not only Linux and Macs but also even Windows which is nice for quick setup of dev environments etc.

Not a deal breaker by any means, but kind of hard to see why adding it would be, either?

**Comment by ylluminarious** [comment]
Well, this is a rather silly argument to have. Why not have different "modes" for the fish compiler? I.e., some sort of setting that you can enable wherein you can have "pure" fish stuff or you can have compatibility with bash-like shells along with fish-isms. You could even have different "levels" of compatibility for other shells so that you can have full compatibility, no compatibility, or something in between. I mean, at a basic level, you could at least implement some sort of source-to-source compiler for this. Or take bash code and translate it to an internal, intermediate representation for fish. You could do all kinds of things for translation in this regard. Correct me if I'm wrong, but I believe that the language-level concepts of bash et al. vs. fish aren't world-shatteringly different, right?

Anyway, that's my $0.02 on this issue that really seems silly because it's built on the premise of:
```
"pure fish"
XOR
"compatibility with bash / becoming another bash clone"
```
when there's nothing that prevents having middle-ground(s) between these two notions.

**Comment by krader1961** [comment]
> ...when there's nothing that prevents having middle-ground(s) between these two notions.

Nothing other than someone (or group of people) investing more than 2,000 hours (a person year) to write the necessary code. I look forward to reviewing pull-requests from you, @ylluminarious, to implement all of that. What you're proposing is equivalent to asking why can't someone make Python also interpret PHP.  The answer is that theoretically you could have a compatibility switch to do so but why would you?

**Comment by ridiculousfish** [comment]
I think we've agreed to implement && and || for 3.0

**Comment by zanchey** [comment]
Implemented in #4620.
